### PR TITLE
Update Psalm to 5.12.0

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement

This patch should unblock the upgrade to PHPUnit 10. `jetbrains/phpstorm-stubs` need to be updated to a development version in order to include the changes from https://github.com/JetBrains/phpstorm-stubs/pull/1531.

It appeared to be quite challenging to implement these changes on top of a 3.x branch. Whoever is interested in doing that can try backporting them.

### CODE REVIEW

File: src/Driver/OCI8/Result.php [hunk 1]
```diff
@@ -73,7 +73,7 @@ public function fetchFirstColumn(): array
         return $this->fetchAll(OCI_NUM, OCI_FETCHSTATEMENT_BY_COLUMN)[0];
     }
 
-    public function rowCount(): int
+    public function rowCount(): int|string
```
Comment (derrabus) [inline src/Driver/OCI8/Result.php]: What's the reason for this change? According to the docs, `oci_num_rows()` returns `int|false`, so we should not encounter a string here. If I recall correctly, the only driver returning row counts as string under certain conditions is MySQLi.

File: src/DriverManager.php [hunk 2]
```diff
@@ -60,8 +62,8 @@
  *     primary?: OverrideParams,
  *     replica?: array<OverrideParams>,
  *     serverVersion?: string,
- *     sharding?: array<string,mixed>,
- *     slaves?: array<OverrideParams>,
+ *     sessionMode?: int,
+ *     url?: string,
```
Comment (derrabus) [inline src/DriverManager.php]: `url` should be gone as per #5850.

**Review by derrabus — CHANGES_REQUESTED** [review]
Thank you. It probably makes sense to leave DBAL 3 on Psalm 4. You might have witnessed me struggling with the upgrade in #5830.

File: src/Driver/OCI8/Result.php (hunk 1 above)
Comment (morozov) [reply src/Driver/OCI8/Result.php]: I changed it consistently with the changes in the interface. But yeah, let me revert the unnecessary changes.

File: src/DriverManager.php (hunk 2 above)
Comment (morozov) [reply src/DriverManager.php]: That's a nice improvement. In this case, I'll need to rework this test: https://github.com/doctrine/dbal/blob/95d093a306e3cd059a5ddf2e461c24c6fddf5a4f/tests/Logging/MiddlewareTest.php#L35-L39

**Comment by morozov** [comment]
> You might have witnessed me struggling with the upgrade in https://github.com/doctrine/dbal/pull/5830.

I only noticed it when I was submitting this PR 🙈

File: src/Driver/OCI8/Result.php (hunk 1 above)
Comment (morozov) [reply src/Driver/OCI8/Result.php]: In the earlier days, when we had some debates about the return types, I made some partial research on how extensions handle the integer overflow. I guess I'll leave my notes here for visibility.

The DBAL returns an `int|string` value from:

1. `Connection::executeStatement()`, the number of rows affected by `INSERT`, `UPDATE` or `DELETE` statement.
2. `Result::rowCount()`, the number of rows returned by a `SELECT` statement.

### MySQLi

We're interested in looking at `mysqli::$affected_rows`, `mysqli_stmt::$affected_rows` and `mysqli_stmt::$num_rows`.

1. `link_affected_rows_read()` ([source](https://github.com/morozov/php-src/blob/aff365871aec54c9a556d7667f131b8638d20194/ext/mysqli/mysqli_prop.c#L186-L190)):

   ```c
   if (rc < ZEND_LONG_MAX) {
	   ZVAL_LONG(retval, (zend_long) rc);
   } else {
	   ZVAL_NEW_STR(retval, strpprintf(0, MYSQLI_LLU_SPEC, rc));
   }
   ```
2. `MYSQLI_RETURN_LONG_INT` ([source](https://github.com/php/php-src/blob/01b3fc03c30c6cb85038250bb5640be3a09c6a32/ext/mysqli/mysqli_priv.h#L82-L90)):

   ```c
   #define MYSQLI_RETURN_LONG_INT(__val) \
   { \
       if ((__val) < ZEND_LONG_MAX) {		\
           RETURN_LONG((zend_long) (__val));		\
       } else {				\
           /* always used with my_ulonglong -> %llu */ \
           RETURN_STR(strpprintf(0, MYSQLI_LLU_SPEC, (__val)));	\
       } \
   }
   ```

This behavior is consistent across all MySQLi use cases: connection/statement, affected rows / number of rows / last insert ID.

### PDO MySQL

`mysql_handle_doer`, which is the handler for `PDO::exec()`, casts [`my_ulonglong`](https://dev.mysql.com/doc/c-api/5.7/en/c-api-data-structures.html#idm45760354552176) (`unsigned long long`) to `int`:

```c
PDO_DBG_RETURN((int)c);
```

Although, interestingly, `pdo_mysql_last_insert_id` uses  `zend_u64_to_str`, i.e. always returns a string.

### SQL Server

We're interested in looking at `sqlsrv_rows_affected()`. Interestingly, `rowCount()` doesn't seem to return the analog of `mysqli_stmt::$num_rows` (e.g. `sqlsrv_num_rows()`) for `SELECT` queries.

Both `sqlsrv_rows_affected()` and `sqlsrv_num_rows()` use `stmt->current_results->row_count()` which is populated via:

```cpp
row_count++;
SQLSRV_ASSERT( row_count < INT_MAX, "Hard maximum of 2 billion rows exceeded in a buffered query" );
```

And returned via `RETURN_LONG`. So the SQL Server drivers can't return a value greater than `PHP_INT_MAX`.

**Comment by greg0ire** [comment]
> It probably makes sense to leave DBAL 3 on Psalm 4.

If we don't, new code we merge up might need a rework to be accepted by Psalm 5. If we do, then I suggest we resort to using the baseline if this is too much work, and then proceed to ignore changes to the baseline while merging up the first time.

File: src/Platforms/AbstractPlatform.php [hunk 3]
```diff
@@ -872,7 +872,7 @@ private function buildCreateTableSQL(Table $table, bool $createForeignKeys): arr
     }
 
     /**
-     * @param list<Table> $tables
+     * @param array<Table> $tables
```
Comment (greg0ire) [inline src/Platforms/AbstractPlatform.php:875]: This is less specific not more, but running this locally, I can see why it is needed: this has ripple effects throughout the codebase.

**Review by greg0ire — APPROVED** [review]

File: src/Platforms/AbstractPlatform.php (hunk 3 above)
Comment (morozov) [reply src/Platforms/AbstractPlatform.php:875]: There are a lot of questionable changes in this PR when it comes to array types, and I won't be able to explain many of them. I can only tell that PHPStan (given the current settings) and Psalm handle them quite differently. For instance, for Psalm `list<T>` is a subtype of `array<int<0,max>,T>` but for PHPStan it's not. So there are a lot of compromises here to satisfy them both.

**Comment by morozov** [comment]
> If we don't, new code we merge up might need a rework to be accepted by Psalm 5. If we do, then I suggest we resort to using the baseline if this is too much work, and then proceed to ignore changes to the baseline while merging up the first time.

I would postpone solving this problem until it becomes a problem.

We could refrain from introducing too much new code in 3.x, so it won't become a problem in the first place. Baselining the existing errors in `3.x` will solve the problem only partially because the new code will be written with the baseline in place but then will be merged to `4.0.x` without the baseline, so it will have to be fixed anyways.

Baselining will only allow us to acknowledge the newly introduced errors but it doesn't mean it will be possible to address them without baselining them as well. Otherwise, it would be possible to implement this patch in `3.x` to begin with.

**Comment by greg0ire** [comment]
Ok, then let's try to keep this in mind while merging up, see if we need to fix Psalm issues at that time and if it is painful to do so.

**Comment by morozov** [comment]
Technically, the changes in all the commits prior to the Psalm upgrade are compatible with the old Psalm version, so they could be backported for the sake of reducing the difference between the DBAL versions. But I don't want to do it myself.

**Review by derrabus — APPROVED** [review]
